ga4gh/avrotools.py

```diff
@@ -152,9 +223,194 @@ def _getInvalidFields(cls, expectedSchema, datum):
             for field in expectedSchema.fields:
                 key = field.name
                 value = datum.get(key)
-                result = cls._getInvalidFields(field.type, value)
-                if result != cls.sinkValue:
+                result = self._getInvalidFields(field.type, value)
+                if result != self.sinkValue:
                     dic[key] = result
             if len(dic):
                 return dic
-        return cls.sinkValue
+        return self.sinkValue
+
+
+class RandomSchemaTool(object):
+    """
+    Generates random instances and values
+    """
+    @classmethod
+    def generateInvalidTypeValue(cls, *typeStrings):
+        """
+        Returns a value that is invalid for the supplied type.
+        """
+        if any(t in cls._getTypeToGenerator() for t in typeStrings):
+            remaining = dict((typeString, gen) for typeString, gen
+                             in cls._getTypeToGenerator().iteritems()
+                             if typeString not in typeStrings)
+            return random.choice(remaining.values())
+        else:  # Not a primitive type, can't be a bool, so just return boolean
+            return cls._getTypeToGenerator()["boolean"]
+
+    @classmethod
+    def _generatorForType(cls, typeString):
+        try:
+            return cls._getTypeToGenerator()[typeString]
+        except KeyError:
+            raise SchemaToolException("unknown type %s" % typeString)
+
+    @classmethod
+    def _getTypeToGenerator(cls):
+        typeToGenerator = {
+            "null": cls._noneGenerator(),
+            "boolean": cls._booleanGenerator(),
+            "int": cls._integerGenerator(),
+            "long": cls._longGenerator(),
+            "string": cls._stringGenerator(),
+            "float": cls._floatGenerator(),
+            "double": cls._doubleGenerator()
+        }
+        return typeToGenerator
+
+    @classmethod
+    def _valueFromSetGenerator(cls, possibilities=[]):
+        """
+        Creates a generator which returns a random selection from the
+        supplied list of possiblities.
+        """
+        while True:
+            yield random.choice(possibilities)
+
+    @classmethod
+    def _noneGenerator(cls):
+        """
+        Creates a generator which always returns None.
+        """
+        while True:
+            yield None
+
+    @classmethod
+    def _booleanGenerator(cls):
+        """
+        Creates a generator which returns only True and False.
+        """
+        gen = cls._valueFromSetGenerator([True, False])
+        while True:
+            yield gen.next()
+
+    @classmethod
+    def _integerGenerator(cls, min=-2 ** 31, max=2 ** 31):
```

Comment:
these default arguments (and others) should be changed to use the avro constants